core/benches/scheduler.rs

```diff
@@ -0,0 +1,494 @@
+                .compute_unit_price;
+
+            // NOTE - setting transaction cost to be `0` for now, so it doesn't bother block_limits
+            // when scheduling.
+            const TEST_TRANSACTION_COST: u64 = 0;
+            if self.container.insert_new_transaction(
+                transaction,
+                MaxAge::MAX,
+                compute_unit_price,
+                TEST_TRANSACTION_COST,
+            ) {
+                unreachable!("test is setup to fill the Container to fullness");
+            }
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct BenchStats {
+    bench_iter_count: usize,
+    num_of_scheduling: usize,
+    // worker reports:
+    num_works: Arc<AtomicUsize>,
+    num_transaction: Arc<AtomicUsize>, // = bench_iter_count * container_capacity
+    tracer_placement: Arc<AtomicUsize>, // > 0
+    // from scheduler().result:
+    num_scheduled: usize, // = num_transaction
+}
+
+impl BenchStats {
+    fn print_and_reset(&mut self) {
+        println!(
+            "per bench stats:
+            number of scheduling: {}
+            number of Works: {}
+            number of transactions scheduled: {}
+            number of transactions per work: {}
+            Tracer placement: {}",
+            self.num_of_scheduling
+                .checked_div(self.bench_iter_count)
+                .unwrap_or(0),
+            self.num_works
+                .load(Ordering::Relaxed)
+                .checked_div(self.bench_iter_count)
+                .unwrap_or(0),
+            self.num_transaction
+                .load(Ordering::Relaxed)
+                .checked_div(self.bench_iter_count)
+                .unwrap_or(0),
+            self.num_transaction
+                .load(Ordering::Relaxed)
+                .checked_div(self.num_works.load(Ordering::Relaxed))
+                .unwrap_or(0),
+            self.tracer_placement.load(Ordering::Relaxed)
+        );
+        self.num_works.swap(0, Ordering::Relaxed);
+        self.num_transaction.swap(0, Ordering::Relaxed);
+        self.tracer_placement.swap(0, Ordering::Relaxed);
+        self.bench_iter_count = 0;
+        self.num_of_scheduling = 0;
+        self.num_scheduled = 0;
+    }
+}
+
+// a bench consumer worker that quickly drain work channel, then send a OK back via completed-work
+// channel
+// NOTE: Avoid creating PingPong within bench iter since joining threads at its eol would
+// introducing variance to bench timing.
+#[allow(dead_code)]
+struct PingPong {
+    threads: Vec<std::thread::JoinHandle<()>>,
+}
+
+impl PingPong {
+    fn new<Tx: TransactionWithMeta + Send + Sync + 'static>(
+        work_receivers: Vec<Receiver<ConsumeWork<Tx>>>,
+        completed_work_sender: Sender<FinishedConsumeWork<Tx>>,
+        num_works: Arc<AtomicUsize>,
+        num_transaction: Arc<AtomicUsize>,
+        tracer_placement: Arc<AtomicUsize>,
+    ) -> Self {
+        let mut threads = Vec::with_capacity(work_receivers.len());
```

Comment:
I think we can just use a single thread here to be as low impact as possible. Then just use [Select](https://docs.rs/crossbeam/latest/crossbeam/channel/struct.Select.html) to select from the vec of receivers.

Reply:
Not sure 3 more threads has meaningful impact on bench here, assume one run bench on a isolated box. I tried to reduce pipeline pressure to Scheduler during bench as much as possible, so to have each thread to pull works, and send back finished-works quickly. Ideally these workers have minimal and const perf.